libraries/state_history/include/eosio/state_history/log_catalog.hpp

```diff
@@ -0,0 +1,261 @@
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/key.hpp>
+
+#include <eosio/chain/types.hpp>
+
+#include <eosio/state_history/log.hpp>
+#include <eosio/state_history/log_config.hpp>
+
+#include <fc/scoped_exit.hpp>
+
+namespace eosio::state_history {
+
+using namespace boost::multi_index;
+
+struct catalogued_log_file {
+   chain::block_num_type            begin_block_num = 0;
+   chain::block_num_type            end_block_num = 0;
+   std::filesystem::path            path_and_basename;  //example: /some/dir/trace-history-50-59   i.e. does NOT include .log nor .index
+   std::optional<state_history_log> log;
+
+   size_t last_used_counter = 0;
+
+   size_t effective_last_used_counter() const {
+      if(!log)
+         return 0;
+      return last_used_counter;
+   }
+
+   catalogued_log_file(const catalogued_log_file&) = delete;
+   catalogued_log_file& operator=(catalogued_log_file&) = delete;
+   catalogued_log_file(chain::block_num_type begin_block_num, chain::block_num_type end_block_num, std::filesystem::path path_and_basename) :
+     begin_block_num(begin_block_num), end_block_num(end_block_num), path_and_basename(path_and_basename) {}
+};
+
+class log_catalog {
+   std::filesystem::path retained_dir;
+   std::filesystem::path archive_dir;
+   uint32_t              max_retained_files = std::numeric_limits<decltype(max_retained_files)>::max();
+   uint32_t              log_rotation_stride = std::numeric_limits<decltype(log_rotation_stride)>::max();
+
+   const state_history_log::non_local_get_block_id_func non_local_get_block_id;
+
+   //cache not just an optimization: when a log file is opened the last block in its log file is used as a determination of the log's end block,
+   // so we don't want to close an old log file while it's being written to during a fork event otherwise we'd effectively corrupt the catalog state
+   struct by_mru {};
+   typedef multi_index_container<
+     catalogued_log_file,
+     indexed_by<
+       ordered_unique<key<&catalogued_log_file::begin_block_num>>,
+       ordered_non_unique<tag<by_mru>,key<&catalogued_log_file::effective_last_used_counter>, std::greater<size_t>>
+     >
+   > catalog_t;
+   catalog_t retained_log_files;
+   std::optional<state_history_log> head_log;
+   const std::filesystem::path head_log_path_and_basename;  //example: /some/dir/trace-history   i.e. does NOT include .log nor .index
+
+   size_t global_used_counter = 0;
+
+public:
+   log_catalog(const log_catalog&) = delete;
+   log_catalog& operator=(log_catalog&) = delete;
+
+   log_catalog(const std::filesystem::path& log_dir, const state_history::state_history_log_config& config, const std::string& log_name,
+               state_history_log::non_local_get_block_id_func non_local_get_block_id = state_history_log::no_non_local_get_block_id_func) :
+     non_local_get_block_id(non_local_get_block_id), head_log_path_and_basename(log_dir / log_name) {
+      std::visit(chain::overloaded {
```

Comment:
We need to do different setup depending on which of the three different types of log configuration is required, so I need three different initialization paths one way or another.

Something I don't like here is the repeated `open_head_log()` -- it'd be nice to either a) do it once outside of the visit, but this isn't possible because the call is different for pruned config; or b) at least do it in a common location inside the visit, such as at the top (for the partitioned log it's in the middle for some reason, I will think on why that's the case).

The "nothing needed" comment is invalid and needs to be removed
https://github.com/AntelopeIO/spring/blob/a29f3f96aa9bfebc3a33eb4c3578bab74bb08527/libraries/state_history/include/eosio/state_history/log_catalog.hpp#L73-L75
(this was an accidental holdover from when `open_head_log()` really was outside the three paths!)

Do you have any more input on what exactly is complicated or potential changes? I think maybe all the code that populates the catalog could be moved in to a separate function (that would keep the 3 visits paths here short and of similar length).

Reply:
I was thinking along the same line of moving the catalog code to a separate function and adding more comments.

Reply:
I split that out in 8c597cd

Reply:
Thanks.